File: ProyectoFInal-Marcelo/backend/internal/services/category_service.go
package services

import (
	"errors"
	"notes-app/internal/dto"
	"notes-app/internal/models"
	"notes-app/internal/repositories"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type CategoryService interface {
	CreateCategory(req dto.CreateCategoryRequest) (*models.Category, error)
	GetCategoryByID(id uuid.UUID) (*models.Category, error)
	GetCategories() ([]models.Category, error)
	UpdateCategory(id uuid.UUID, req dto.UpdateCategoryRequest) (*models.Category, error)
	DeleteCategory(id uuid.UUID) error
}

type categoryService struct {
	categoryRepo repositories.CategoryRepository
}

func NewCategoryService(categoryRepo repositories.CategoryRepository) CategoryService {
	return &categoryService{
		categoryRepo: categoryRepo,
	}
}

func (s *categoryService) CreateCategory(req dto.CreateCategoryRequest) (*models.Category, error) {
	// Check if category with same name exists
	existingCategory, err := s.categoryRepo.GetByName(req.Name)
	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}
	if existingCategory != nil {
		return nil, errors.New("category with this name already exists")
	}

	category := &models.Category{
		Name:  req.Name,
		Color: req.Color,
	}

	// Set default color if not provided
	if category.Color == "" {
		category.Color = "#6366f1"
	}

	if err := s.categoryRepo.Create(category); err != nil {
		return nil, err
	}

	return category, nil
}

func (s *categoryService) GetCategoryByID(id uuid.UUID) (*models.Category, error) {
	category, err := s.categoryRepo.GetByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("category not found")
		}
		return nil, err
	}
	return category, nil
}

func (s *categoryService) GetCategories() ([]models.Category, error) {
	return s.categoryRepo.GetAll()
}

func (s *categoryService) UpdateCategory(id uuid.UUID, req dto.UpdateCategoryRequest) (*models.Category, error) {
	// Get existing category
	category, err := s.categoryRepo.GetByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("category not found")
		}
		return nil, err
	}

	// Check if new name already exists (if name is being updated)
	if req.Name != "" && req.Name != category.Name {
		existingCategory, err := s.categoryRepo.GetByName(req.Name)
		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, err
		}
		if existingCategory != nil {
			return nil, errors.New("category with this name already exists")
		}
		category.Name = req.Name
	}

	// Update color if provided
	if req.Color != "" {
		category.Color = req.Color
	}

	if err := s.categoryRepo.Update(category); err != nil {
		return nil, err
	}

	return category, nil
}

func (s *categoryService) DeleteCategory(id uuid.UUID) error {
	// Check if category exists
	_, err := s.categoryRepo.GetByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("category not found")
		}
		return err
	}

	return s.categoryRepo.Delete(id)
}

File: ProyectoFInal-Marcelo/frontend/src/services/categoryService.ts
import api from './api';
import { Category } from '../types/note.types';
import { ApiResponse } from '../types/api.types';

export interface CreateCategoryRequest {
  name: string;
  color?: string;
}

export interface UpdateCategoryRequest {
  name?: string;
  color?: string;
}

export const categoryService = {
  // Create a new category
  createCategory: async (data: CreateCategoryRequest): Promise<Category> => {
    const response = await api.post<ApiResponse<Category>>('/categories', data);
    return response.data.data!;
  },

  // Get all categories
  getCategories: async (): Promise<Category[]> => {
    const response = await api.get<ApiResponse<Category[]>>('/categories');
    return response.data.data!;
  },

  // Get a specific category by ID
  getCategory: async (id: string): Promise<Category> => {
    const response = await api.get<ApiResponse<Category>>(`/categories/${id}`);
    return response.data.data!;
  },

  // Update a category
  updateCategory: async (id: string, data: UpdateCategoryRequest): Promise<Category> => {
    const response = await api.put<ApiResponse<Category>>(`/categories/${id}`, data);
    return response.data.data!;
  },

  // Delete a category
  deleteCategory: async (id: string): Promise<void> => {
    await api.delete(`/categories/${id}`);
  },
};